File: apps/mobile/lib/services/drive_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:logger/logger.dart';
import 'package:mobile/data/client/lib/api.dart';
import 'package:mobile/services/token_bearer_http_client.dart';
import 'package:mobile/services/token_storage.dart';

final _logger = Logger();

/// Google Drive ファイルモデル
// TODO: OpenAPIの自動生成モデルを使うように変更
class DriveFile {
  final String id;
  final String name;
  final String? modifiedTime;
  final bool isVideo;
  final String? thumbnailLink;
  final String? webViewLink;

  DriveFile({
    required this.id,
    required this.name,
    this.modifiedTime,
    required this.isVideo,
    this.thumbnailLink,
    this.webViewLink,
  });
}

/// Google Drive エラー
class DriveException implements Exception {
  final String message;
  final int? statusCode;

  DriveException(this.message, [this.statusCode]);

  @override
  String toString() => 'DriveException: $message (status: $statusCode)';
}

/// Google Drive サービス
///
/// OpenAPI 自動生成クライアントを使用
/// TokenBearerHttpClient で JWT トークンを自動付与
class DriveService {
  late final ApiClient _apiClient;
  late final GDriveApi _gDriveApi;
  late final TokenStorage _tokenStorage;

  DriveService() {
    // TokenStorage 初期化
    _tokenStorage = TokenStorage();

    // OpenAPI クライアント初期化
    final basePath = _getServerBasePath();
    _logger.i('DriveService: basePath=$basePath');

    // TokenBearerHttpClient で JWT トークンを自動付与
    final httpClient = TokenBearerHttpClient(_tokenStorage);
    _apiClient = ApiClient(basePath: basePath);
    _apiClient.client = httpClient;
    _apiClient.addDefaultHeader('X-Client-Type', 'flutter');

    // GDrive API 初期化
    _gDriveApi = GDriveApi(_apiClient);
  }

  /// プラットフォームに応じてサーバーの basePath を取得
  String _getServerBasePath() {
    return 'http://api.danmaku.cloud:3001';
  }

  /// Google Drive フォルダ内のファイル一覧を取得
  ///
  /// GET /api/gdrive/list
  /// @param folderId フォルダID（デフォルト: 'root'）
  /// @return DriveFile のリスト
  Future<List<DriveFile>> listFolder({String folderId = 'root'}) async {
    try {
      _logger.i('DriveService: フォルダ一覧を取得中 (folderId=$folderId)');

      // OpenAPI で GET /api/gdrive/list を呼び出し
      final response = await _gDriveApi.gDriveControllerListFolderWithHttpInfo(
        folderId,
      );

      // ステータスコード確認
      if (response.statusCode == null || response.statusCode! >= 400) {
        throw DriveException(
          'Failed to list folder',
          response.statusCode,
        );
      }

      // レスポンスボディをJSON デコード
      if (response.body.isEmpty) {
        throw DriveException('No data received from list folder endpoint');
      }

      final data = jsonDecode(response.body);
      _logger.i('DriveService: レスポンス受信: $data');

      // items の配列をマッピング
      if (data is! Map<String, dynamic> || data['items'] is! List) {
        throw DriveException('Invalid response format from server');
      }

      final items = (data['items'] as List)
          .map((item) => _mapFileItemDtoToDriveFile(item))
          .toList();

      _logger.i('DriveService: フォルダ一覧取得成功: ${items.length} 個');
      return items;
    } catch (e) {
      _logger.e('DriveService: フォルダ一覧取得失敗', error: e);
      rethrow;
    }
  }

  /// Google Drive でキーワード検索
  ///
  /// GET /api/gdrive/search
  /// @param folderId 検索対象フォルダID
  /// @param query 検索キーワード
  /// @return DriveFile のリスト
  Future<List<DriveFile>> search({
    required String folderId,
    required String query,
  }) async {
    try {
      _logger.i(
        'DriveService: 検索実行中 (folderId=$folderId, query=$query)',
      );

      // GET /api/gdrive/search を呼び出し
      final response = await _gDriveApi.gDriveControllerSearchWithHttpInfo(
        folderId,
        query,
      );

      // ステータスコード確認
      if (response.statusCode == null || response.statusCode! >= 400) {
        throw DriveException(
          'Failed to search in folder',
          response.statusCode,
        );
      }

      // レスポンスボディをJSON デコード
      if (response.body.isEmpty) {
        throw DriveException('No data received from search endpoint');
      }

      final data = jsonDecode(response.body);
      _logger.i('DriveService: 検索レスポンス受信: $data');

      // items の配列をマッピング
      if (data is! Map<String, dynamic> || data['items'] is! List) {
        throw DriveException('Invalid response format from server');
      }

      final items = (data['items'] as List)
          .map((item) => _mapFileItemDtoToDriveFile(item))
          .toList();

      _logger.i('DriveService: 検索完了: ${items.length} 件');
      return items;
    } catch (e) {
      _logger.e('DriveService: 検索失敗', error: e);
      rethrow;
    }
  }

  /// FileItemDto を DriveFile にマッピング
  DriveFile _mapFileItemDtoToDriveFile(dynamic item) {
    final id = item['id'] as String;
    final name = item['name'] as String;
    final mimeType = item['mimeType'] as String;
    final modifiedTime = item['modifiedTime'] as String?;
    final thumbnailLink = item['thumbnailLink'] as String?;
    final webViewLink = item['webViewLink'] as String?;

    // フォルダかビデオかを判定
    final isVideo = mimeType.contains('video');

    return DriveFile(
      id: id,
      name: name,
      modifiedTime: modifiedTime,
      isVideo: isVideo,
      thumbnailLink: thumbnailLink,
      webViewLink: webViewLink,
    );
  }
}

File: apps/mobile/src/services/drive-service.ts
/**
 * フォルダ・ドライブサービス
 * OpenAPI 自動生成クライアントを使用
 * マルチプロバイダー対応（connectionId が必須）
 */

import { appLogger } from '@/utils/logger';
import { FileItemDto, FolderListDto } from '@/generated';
import { FolderApi } from '@/generated';
import { createApiConfiguration } from './api-config';

export class DriveException extends Error {
  constructor(
    message: string,
    public statusCode?: number
  ) {
    super(`DriveException: ${message} (status: ${statusCode})`);
  }
}

export class DriveService {
  private folderApi: FolderApi;
  private token?: string;

  constructor(token?: string) {
    try {
      this.token = token;
      // OpenAPI Configuration を設定（ドライブ固有のトークンを渡す）
      const config = createApiConfiguration(token);
      this.folderApi = new FolderApi(config);
      appLogger.info('DriveService: 初期化完了');
    } catch (error) {
      appLogger.error('DriveService: 初期化失敗', error);
      throw error;
    }
  }

  /**
   * 特定の接続からフォルダ内容を取得（マルチプロバイダー対応）
   * GET /api/drive/connections/:connectionId/files
   */
  async listFolderByConnection(
    connectionId: string,
    folderId?: string
  ): Promise<FolderListDto> {
    try {
      appLogger.info(
        `DriveService: 接続別フォルダ一覧を取得中 (connectionId=${connectionId}, folderId=${folderId})`
      );

      const response = await this.folderApi.folderControllerListFolderByConnection(
        {
          connectionId,
          folderId,
        }
      );

      appLogger.info(
        `DriveService: 接続別フォルダ一覧取得成功: ${response?.items?.length || 0} 個`
      );

      return response;
    } catch (error) {
      appLogger.error('DriveService: 接続別フォルダ一覧取得失敗', error);
      throw new DriveException(
        'Failed to list folder by connection',
        (error as any)?.status
      );
    }
  }

  /**
   * 接続別検索
   * GET /api/drive/connections/:connectionId/search
   */
  async searchByConnection(
    connectionId: string,
    folderId: string,
    query: string
  ): Promise<FolderListDto> {
    try {
      appLogger.info(
        `DriveService: 接続別検索実行中 (connectionId=${connectionId}, folderId=${folderId}, query=${query})`
      );

      const response = await this.folderApi.folderControllerSearchByConnection({
        connectionId,
        folderId,
        query,
      });

      appLogger.info(
        `DriveService: 接続別検索完了: ${response?.items?.length || 0} 件`
      );

      return response;
    } catch (error) {
      appLogger.error('DriveService: 接続別検索失敗', error);
      throw new DriveException(
        'Search by connection failed',
        (error as any)?.status
      );
    }
  }
}

/**
 * DriveService ファクトリ
 * 複数ドライブ対応のため、トークンごとに新しいインスタンスを生成
 */
export const createDriveService = (token?: string): DriveService => {
  return new DriveService(token);
};

// デフォルト（auth-store のトークン使用）
export const driveService = new DriveService();
